cpp/src/arrow/array/builder_dict.h

```diff
@@ -282,6 +294,163 @@ class DictionaryBuilderBase : public ArrayBuilder {
     return indices_builder_.AppendEmptyValues(length);
   }
 
+  Status AppendScalar(const Scalar& scalar, int64_t n_repeats) override {
+    if (!scalar.type->Equals(type())) {
+      return Status::Invalid("Cannot append scalar of type ", scalar.type->ToString(),
+                             " to builder for type ", type()->ToString());
+    }
+    if (!scalar.is_valid) return AppendNulls(n_repeats);
+
+    const auto& dict_ty = internal::checked_cast<const DictionaryType&>(*scalar.type);
+    const DictionaryScalar& dict_scalar =
+        internal::checked_cast<const DictionaryScalar&>(scalar);
+    const auto& dict = internal::checked_cast<const typename TypeTraits<T>::ArrayType&>(
+        *dict_scalar.value.dictionary);
+    switch (dict_ty.index_type()->id()) {
+      case Type::UINT8: {
+        const auto& value = dict.GetView(
+            internal::checked_cast<const UInt8Scalar&>(*dict_scalar.value.index).value);
```

Comment:
What happens if `dict` has a null at this index?

Reply:
Hmm, there should be better testing for nulls in general, I'll amend that.